### DISCUSSION TITLE
Indoor navigation using laser tracker instead of GPS

### DISCUSSION THREAD

**Discussion by JanGrizz** [iNavFlight/inav#8484] [Q&A]
Hi folks,
I´m new here and I need help with my project. I´m trying to use a laser tracker for position feedback on a microquad (later larger drones).

As part of my master thesis I´m working on a indoor navigation system for drones. The idea is to navigate a done (any) in a non GPS environment with high precison. I´m using a laser tracker "Leica AT901" to get the current position of the drone. The use of this laser tracker is mandatory. Later there will be cameras on a drone to take pictures. The pictures and position data will be used to generate precise 3D models.
First step would be to just hold the position and later on, the drone has to navigate to a couple of waypoints.
My first atempt was to convert my cartesian measurements from the laser tracker to a fake-GPS signal (NMEA-protocol) and send it via RF (433 MHz) to the drone. The conversion is realized on a PC via a Python script.
The receiver is connected to one of the UART ports. The port ist configured for a GPS.
When I connect my M8N-GPS with compass and barometer (I added the baro into the housing) the FC recognizes all sensors and works fine. When I change to my synthetic fake-GPS signal, which I created according to the output of the M8N-GPS sensor and NMEA standards, the FC does not recognize the fake-GPS signal reliable. It jumps randomly between "GPS = OK" and "GPS = not available" indicated by the blue/red GPS icon, Hardware health OK/not OK and the status through CLI tells the same story.
I used an Arduino Uno to observe the communication between FC to M8N-GPS and FC to RF-serial (fake-GPS). The observed messages match my expectations and should work (I think so at least).
Plugging in a battery does not solve the problem (on screenshot is only USB power).

I choose this approach because I wanted to be able to just plug a RF receiver into any drone (UART) and have it ready to navigate with a laser tracker. Now I guess I have to move to a different approach...

My questios: Is my GPS-fake-attempt a good/solvable idea? If so how do I get my FC to accept my signal, what am I missing? How else could I use my cartesian measurements to navigate a drone using INAV?

I wanted to avoid to mess with the firmware of the FC, but I guess I have to. I have mediocre programming skills in C++ and Python. I am willing to improove my skills, but I need some help!

Regards
Jan

My setup:
INAV firmware: 5.1.0
FC: Mamba F722S Stack
GPS: M8N with compass HCM5883L
BARO: BMP280
RF-UART: SV611 (433MHz)
Quad frame: Iflight Cidora

This is what I get from the laser trcker [x y z]:
```
[3039.31992256 -364.8647846  -285.44531833]
```

This is the message of the M8N GPS (one block) I used as a template (the position is not accurate);
```
$GPRMC,100056.00,A,5225.55201,N,00936.91729,E,0.026,,180822,,,A*7C
$GPVTG,,T,,M,0.026,N,0.048,K,A*2B
$GPGGA,100056.00,5225.55201,N,00936.91729,E,1,10,0.84,85.3,M,46.0,M,,*64
$GPGSA,A,3,01,21,03,27,32,14,22,08,10,17,,,1.57,0.84,1.33*03
$GPGSV,3,1,12,01,51,285,49,03,22,226,34,08,54,182,41,10,27,056,27*78
$GPGSV,3,2,12,14,24,300,39,17,07,316,34,21,81,295,33,22,35,121,45*72
$GPGSV,3,3,12,24,04,021,26,27,22,156,32,28,16,325,30,32,41,089,37*72
$GPGLL,5225.55201,N,00936.91729,E,100056.00,A,A*60
```

This is the message (one block) I create in Python and send to FC:
``` 
$GPRMC,140225.00,A,5242.59450,N,00961.52290,E,0.026,,201022,,,A*79
$GPVTG,,T,,M,0.026,N,0.048,K,A*2B
$GPGGA,140225.00,5242.59450,N,00961.52290,E,1,10,0.84,46.0,M,46.0,M,,*6F
$GPGSA,A,3,01,21,03,27,32,14,22,08,10,17,,,1.57,0.84,1.33*3
$GPGSV,3,1,12,01,51,285,49,03,22,226,34,08,54,182,41,10,27,056,27*78
$GPGSV,3,2,12,14,24,300,39,17,07,316,34,21,81,295,33,22,35,121,45*72
$GPGSV,3,3,12,24,04,021,26,27,22,156,32,28,16,325,30,32,41,089,37*72
$GPGLL,5225.55201,N,00936.91729,E,100056.00,A,A*60
```

These are the GPS settings Iused on the FC:
```
set gps_provider = NMEA
set gps_sbas_mode = NONE
set gps_dyn_model = AIR_1G
set gps_auto_config = OFF
set gps_auto_baud = OFF
set gps_ublox_use_galileo = OFF
set gps_min_sats = 6
```

Code on Arduino Uno:
```
#include <SoftwareSerial.h>

#define PIN_RXD 3
#define PIN_TXD 4

#define GPS_BAUD 9600
#define SERIAL_BAUD 9600

// serial connection to the GPS device
SoftwareSerial gpsSerial(PIN_RXD, PIN_TXD);

void setup(){
    Serial.begin(SERIAL_BAUD);
    gpsSerial.begin(GPS_BAUD);
}

void loop(){
  // Output raw GPS data to the serial monitor
  while (gpsSerial.available() > 0) {
    Serial.write(gpsSerial.read());
  }
  while (Serial.available() > 0) {
    gpsSerial.write(Serial.read());
  }
}
```

![INAV Screenshot Ports](https://user-images.githubusercontent.com/116190297/196980998-f9156a12-033a-467b-ad39-b0042d2d5183.png)
![INAV Screenshot Setup](https://user-images.githubusercontent.com/116190297/196981004-cb8665b7-ec04-4960-8f55-e7cb69cf762d.png)

**Comment by stronnag** [comment]
The GPS component in INAV 5.1 (and earlier) requires a  less than 1 second update rate, otherwise it will fail the GPS update check.
This is fixed for 6.0 in #8423, however be aware that NMEA will be deprecated in 6.0 and removed in probably 7.0.

**Reply by JanGrizz** [reply]
Well that was an unexpected simple fix...
I´m a lttle bit embarressed by myself!

Well I better get working on a UBOX emulator then...

Thank you so much!
